clang_delta/ReplaceArrayAccessWithIndex.cpp

```diff
@@ -0,0 +1,137 @@
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif
+
+#include "ReplaceArrayAccessWithIndex.h"
+
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/Basic/SourceManager.h"
+
+#include "TransformationManager.h"
+
+#include <iostream>
+
+using namespace clang;
+using namespace llvm;
+
+
+static const char *Description =
+  "Replace array accesses with the index expression.";
+
+static RegisterTransformation<ReplaceArrayAccessWithIndex>
+Trans("replace-array-access-with-index", Description);
+
+struct ReplaceArrayAccessWithIndex::IndexCollector
+  : public RecursiveASTVisitor<ReplaceArrayAccessWithIndex::IndexCollector>
+{
+  IndexCollector(ReplaceArrayAccessWithIndex *instance);
+  bool VisitArraySubscriptExpr(ArraySubscriptExpr *ASE);
+
+  const VarDecl *getVarDeclFromExpr(const Expr *E);
+
+  ReplaceArrayAccessWithIndex *ConsumerInstance;
+};
+
+ReplaceArrayAccessWithIndex::IndexCollector::IndexCollector(
+  ReplaceArrayAccessWithIndex *instance)
+  : ConsumerInstance(instance)
+{
+  // No further initialization needed.
+}
+
+bool ReplaceArrayAccessWithIndex::IndexCollector::VisitArraySubscriptExpr(
+  ArraySubscriptExpr *ASE)
+{
+  // Skip expressions in included files.
+  if (ConsumerInstance->isInIncludedFile(ASE))
+    return true;
+
+  // Skip multi-dimensional arrays.
+  const VarDecl *BaseVD = getVarDeclFromExpr(ASE->getBase());
+
+  if (!BaseVD)
+    return true;
+
+  ArrayType const *ArrayTy = dyn_cast<ArrayType>(BaseVD->getType().getTypePtr());
+
+  if (!ArrayTy || dyn_cast<ArrayType>(ArrayTy->getElementType().getTypePtr()))
```

Comment:
Can we also skip cases where the array element type is incompatible with integer type? For example, with the current implementation, this transformation turns the code below

    struct S { int f; };
    int foo() {
      S s[1] = {{1}};
      return s[0].f;
    }

into the following form that can't be compiled:

    struct S { int f; };
       int foo() {
      S s[1] = {{1}};
      return (struct S)(0).f;
    }

Adding more restrictions can reduce the chances of generating "bad" output.